Add DataTable

fejes713 opened the pull request:
### This is probably really awful snippet but let's learn from mistakes as we did in 30code.

_TODO:_
- Description needs improvement definitely
- Typos are probably present, fix them
- Code could be prettier
- Somehow let user know he can use custom table cells as another component

Chalarangelo commented on snippets/mapped-table.md:
```diff
@@ -0,0 +1,35 @@
+function MappedTable(props) {
+  return (
+    <tbody>
+      {props.headers ? return(<tr>{props.headers.map((h, i)=> (<th key={i}>h</th>))}</tr>) : return null}
```
The `return` keywords are not necessary in either `{}` set, there are shorthands for those.

Chalarangelo commented on snippets/mapped-table.md:
````diff
@@ -0,0 +1,35 @@
+```jsx
+function MappedTable(props) {
+  return (
+    <tbody>
````
The `<tbody>` could be moved further down to render only the body and the whole `headers` part could be wrapping a `<thead>` element to be rendered on condition - this is more semantically correct.

Chalarangelo commented on snippets/mapped-table.md:
```diff
@@ -0,0 +1,35 @@
+          </tr>
+        );
+      })}
+    </tbody>
```
Technically speaking, if we want to be very strict about semantics, we should allow users to pass `props.footers` to actually add a footer to the table and `props.caption` to caption it. If that is necessary or if it's a common use-case remains to be seen, I'm just stating this to be on the record for further discussion.

Chalarangelo commented on snippets/mapped-table.md:
````diff
@@ -0,0 +1,35 @@
+);
+```
+
+<!-- tags: (functional,array) -->
````
In my head, the tagger implementation that will generate the TOC will use the first tag as primary category, so I'd rather we had the order of these reversed, but that is just a suggestion as we have no script implementations set up yet.

Chalarangelo commented on snippets/mapped-table.md:
```diff
@@ -0,0 +1,35 @@
+
+<!-- tags: (functional,array) -->
+
+<!-- expertise: (0) -->
```
I tagged #8 as `1` (intermediate), not certain about what should be what, so we should let visitors and the rest of the team suggest the expertise for both to get a feel for them.

Chalarangelo commented on snippets/mapped-table.md:
```diff
@@ -0,0 +1,35 @@
+### MappedTable
+
+Generate a table with dynamic rows based on passed array of data. Optionally include headers.
```
How about:

*Renders a table with rows dynamically created from an array of data.*

And then in the below explanation, add a line that states *Omit `headers`... blah blah*, like we usually do in 30code?

Chalarangelo commented on snippets/mapped-table.md:
```diff
@@ -0,0 +1,35 @@
+
+Generate a table with dynamic rows based on passed array of data. Optionally include headers.
+
+Conditionaly check weather to include table headers or not. Use `Array.prototype.map` to map over array of data and return `<tr>` for each entry.
```
Change lines on fullstops to make maintaining slightly easier (this helps us have cleaner diffs like we do in other projects).
Read the above suggestion about omitting arguments as well.
What `data` is should probably be explained as well - see [here](https://github.com/30-seconds/30-seconds-of-react/pull/8/files#diff-bcaa462389dee12dde86b1542c6fd40dR7) for my suggestion.

fejes713 commented:
- Updated snippet to include better and more accurate description. Omitted return keywords where possible.
- Updated semantics so table now consists of `table`, `thead` and `tbody` elements. 
- Changed order of tags and updated difficulty to 1.

I am still not sure about footer as it would increase the snippet without too much value. Just same code as in the header. I am open for discussion on this one.

atomiks commented on snippets/mapped-table.md:
```diff
@@ -0,0 +1,45 @@
+      </thead>
+      <tbody>
+        {props.data.map((v, i) => {
+          return (
```
Omit the return keyword?

atomiks commented on snippets/mapped-table.md:
```diff
@@ -0,0 +1,45 @@
+  return (
+    <table>
+      <thead>
+        {props.headers ? (
```
Maybe `props.header` if there's only one

fejes713 replied:
Yep, agree with this one

atomiks commented on snippets/mapped-table.md:
````diff
@@ -0,0 +1,45 @@
+
+```jsx
+const people = [{ id: 841, name: 'John' }, { id: 241, name: 'Jesse' }];
+const headers = ['id', 'name'];
````
Maybe we can use `Object.keys()` on the `people` object instead of specifying `headers` separately

fejes713 commented:
Thanks @meetzaveri 👍 I'll fix this today. And yes @skatcat31 there's 1 extra header, I'll have to see how to generalize this function a bit 😢 

The general idea is that people won't render classic `<td>` elements but instead they will use their custom components. (I did something similar on my internship). Because of this, I agree that snippets should be generalized as much as possible so they can be applied in many cases.

nikolalsvk commented on snippets/mapped-table.md:
````diff
@@ -0,0 +1,45 @@
+Use the value of the `headers` prop to conditionally render a table with a header. Use `Array.prototype.map` to render every item in `data` as a `<tr>` element and give it a `key`. `data` can either be an array of objects with the `id` and `value` properties or an array of primitives. Omit the `headers` prop to render a `table` without headers.
+
+```jsx
+function MappedTable(props) {
````
A lot of people pass props like this:

```js
function MappedTable({ headers, data }) {
```

They destructure the props in the function definition (or functional component definition). This is serves kinda as a documentation since you see all the props that need to be passed to that function/component :D

fejes713 replied:
Makes total sense to me. Much more readable in every aspect 👍 

@Chalarangelo @skatcat31 could you guys give your opinion on this one? I think we should enforce this rule.  I vote +1 for this for the following reasons:

- Very readable
- Users don't need to look into component's code in order to see what props are used
- Eliminates looking into examples of obvious components/props 👍 

Thanks @nikolalsvk

Chalarangelo replied:
I agree, this is a good idea and will help a lot in the long run.

fejes713 commented:
I've updated my example to include changes proposed by @meetzaveri and @nikolalsvk. 

There's another problem that I don't happen to know how to solve. @skatcat31 mentioned that snippets should be generalized as much as possible and **I totally agree with that**. 

This example renders a table with columns `id` and `name`, any idea how to generalize this a bit?

```jsx
            <tr obj={v} key={v.id ? v.id : i}>
              <td>v.id</td>
              <td>v.name</td>
            </tr>
```

meetzaveri commented:
You can use logical or operator like `v.id || i`. This will evaluate `v.id` if there or `i` if not. 

So simple generalization on this would be the conditional approach - `v.id || i`

fejes713 commented:
> You can use logical or operator like `v.id || i`. This will evaluate `v.id` if there or `i` if not.
> 
> So simple generalization on this would be the conditional approach - `v.id || i`

I am really stupid 😢

nikolalsvk commented:
I'm sorry, but I'll have to disagree with using index as a React element key:

```js
// bad
<tr obj={v} key={v.id ? v.id : i}>
  <td>v.id</td>
  <td>v.name</td>
</tr>

// good
<tr obj={v} key={v.id}>
  <td>v.id</td>
  <td>v.name</td>
</tr>
```

There are couple of reasons why you should **NEVER** use index as the key of a React element. This article describes it nicely https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318
Also, great example is given in this stackoverflow answer https://stackoverflow.com/a/42775667/4158895

Chalarangelo commented:
I will have to make similar changes to the footprint of #8. 

Meanwhile, as far as keys are concerned, it's obviously preferred to supply `id`s for them, but if the user doesn't supply anything, **there has to be a fallback**. The snippet makes it clear that people should pass their own `id`s, instead of using indexes and I think this is a good enough way to implement this. Having to generate unique keys from the data in any other way would probably make the snippet too verbose.

meetzaveri commented:
> I will have to make similar changes to the footprint of #8.
> 
> Meanwhile, as far as keys are concerned, it's obviously preferred to supply `id`s for them, but if the user doesn't supply anything, **there has to be a fallback**. The snippet makes it clear that people should pass their own `id`s, instead of using indexes and I think this is a good enough way to implement this. Having to generate unique keys from the data in any other way would probably make the snippet too verbose.

Yes @Chalarangelo  fallback is crucial validation in any scenario. That's why we always use `index` as fallback

skatcat31 commented:
Suggestion on generalizing: the snippet needs to do the most general part of the module and let the user supply the integral part. You don't want to format their divs, you only want to get rid of the loop they would have to do otherwise and the boiler plate. individual formatting should be a conscious developer decision because data is unique to each use case. Suggested change with formatter and optional header formatter.

Also I changed the case on the function name because it's not a type/class, which are often the only reason to title case something in programming, but that can be discussed in styling

```jsx
function mappedTable({ data, headers, format }) {
  return (
    <table>
        {headers && (
          <thead>
            {headers}
          </thead>
        )}
      <tbody>
        {data.map(format}</tr>)}
      </tbody>
    </table>
  );
}
```

example:

```jsx
const people = [{ id: 841, name: 'John' }, { id: 241, name: 'Jesse' }];
ReactDOM.render(
  <mappedTable data={people}
    headers = {Object.keys(data[0]).map(key => <th key={key}>{key}</th>})}
    format = {dataPoint => <tr key={datapoint.id}><td>ID: {datapoint.id}</td><td>Name: {dataPoint.name}</td></tr>}
  />,
  document.getElementById('root')
);
```

Now when writing it reads like the following story for the developer:

"I want a mapped table displaying headers of this specific format, and the data formatted this way through map"

The developer no longer has to go about the boilerplate and can create semantic code. That's a win.

Ideally however I'd probably move this to using children for the formatter:
(warning: I have not tested this due to time constraints. I might have borked the logical `||` for onlyChild condition on passing the `formatter` or possibly how I'm de-referencing them)

```jsx
function mappedTable({ data, headers, format }) {
  let onlyChild;
  try{
    onlyChild = React.Children.only(props.children);
  }catch(_){
    onlyChild = false;
  }
  return (
    <table>
        {!onlyChild && (
          <thead>
            {props.children[0]}
          </thead>
        )}
      <tbody>
        {data.map(onlyChild || props.children[1])}
      </tbody>
    </table>
  );
}
```

```jsx
const people = [{ id: 841, name: 'John' }, { id: 241, name: 'Jesse' }];

ReactDOM.render(
  <mappedTable data={people}>
    {Object.keys(data[0]).map(key => <th key={key}>{key}</th>})}
    {dataPoint => <tr key={datapoint.id}><td>ID: {datapoint.id}</td><td>Name: {dataPoint.name}</td></tr>}
  </mappedTable>,
  document.getElementById('root')
);
```

Although now it begs the question of "should we do the boilerplate for `<thead>` && `</thead>` or should we force the developer to do that too?

```jsx
function mappedTable({ data, headers, format }) {
  let onlyChild;
  try{
    onlyChild = React.Children.only(props.children);
  }catch(_){
    onlyChild = false;
  }
  return (
    <table>
      {!onlyChild && props.children[0]}
      <tbody>
        {data.map(onlyChild || props.children[1])}
      </tbody>
    </table>
  );
}
```

```jsx
const people = [{ id: 841, name: 'John' }, { id: 241, name: 'Jesse' }];

ReactDOM.render(
  <mappedTable data={people}>
    <thead>{Object.keys(data[0]).map(key => <th key={key}>{key}</th>})}</thead>
    {dataPoint => <tr key={datapoint.id}><td>ID: {datapoint.id}</td><td>Name: {dataPoint.name}</td></tr>}
  </mappedTable>,
  document.getElementById('root')
);
```

Chalarangelo commented:
@skatcat31 From a learning standpoint, this is a very verbose snippet and requires a lot of work on the developer's side. A semi-opinionated component that does some stuff by itself is a lot easier to read and understand, let alone tweak and hack to your liking. 

30-seconds projects have always been about teaching by example, while maintaining a balance for simple use-cases. I think that creating a simple table from a data source is just about enough. Same applies to #8, which is a simple list from data. These snippets are targeted at beginners mostly and do not aim to provide a full-fledged component that could be 100+ lines of code. 

Meanwhile, I am implementing a collapse component for my next PR. I will try to make it as unopinionated as possible, but in general we have to do some CSS-in-JS to make these work and we have to make components as simple as possible with the least code. There are hundreds of libraries that create lot more robust components, we are not trying to compete with them. We are trying to demistify the ones that people use and give them a decent starting point to implement something they can use.

skatcat31 commented:
@Chalarangelo React is an advanced topic in and of itself would be an argument to be made, but that's beside the point 👺(tengu used here to denote personal feeling above what could be considered true)

However `30react.mappedList()` is a much more simple concept since lists are a more simple concept. Tables are always verbose because you have more data to display. I would still say it should have a formatter however. If you're using react and don't understand how that a function can be passed...

This is however an introduction to `Children` and `Children.only()`. Maybe this one should be an advanced snippet?

If not we could have an archive version of a generic `mappedTable` so that people who want to click the link in the non advanced version can see how it can be expanded to a more advanced version that can be used in more places with custom formatting at each instance and save them boilerplate?

skatcat31 commented:
@Chalarangelo another suggestion:

`mappedList`, `basicList`, `mappedTable`, `basicTable` may be four snippets that could all show similar, yet different approaches to the same situation(or the basics can link to an archive advanced version with mapping)

I think at this point I'm getting hung up that we're calling them `mapped` yet have nothing to `map`?

Chalarangelo commented:
`Mapped` referred to mapping `data` via `Array.map` to a view. `BasicTable` and `BasicList` sound a bit funny to me, honestly, but maybe the naming doesn't do the snippets justice. 

We certainly need more advanced versions of snippets, which could be like evolutions of snippets (possibly a very neat website feature where you can see if a snippet has a supercharged version and view it if you want?). However, that early in the project's development, we should stick to very basic implementations to get a feel for everything and then expand onto more advanced ones. Slow and steady wins the race! 😉

skatcat31 commented:
so then archived more advanced suggestion it is then(advanced due to concepts in play, not length).

Chalarangelo commented:
**Summary:** More advanced versions should be implemented, but not until we have some guidelines and some feedback. Naming for this and #8 seems to be a bit weird, so we kinda need to come up with better names for the basic versions of these snippets. @fejes713 (if you do not have much time to read through the whole thing)

Chalarangelo commented:
@skatcat31 you know you could just name them `DataTable` and `DataList`, that sounds reasonably simple and it literally describes the component - a table/list created from some data.

skatcat31 commented:
@Chalarangelo the basic ones or the advanced ones as `DataXXXXX`?

Chalarangelo commented:
@skatcat31 basic ones as `DataSomething`, advanced ones as `MappedSomething` (or similar).

Chalarangelo commented:
I'm going to pick this up in a bit, switch it to `DataTable` and apply the following changes:

* Use only one prop, `data`, a normal array of values.
* Render the headers as `id` and `data`.
* Simplify code to work with said changes.

Revisiting this, I realize we should make this a very simple example of printing out a table from an array of values, which has its use-cases. 

I will also PR a second snippet, `MappedTable` that will render an array of objects, based on a second prop, `headers`, which will filter the properties of said objects and only render those. This would be a more advanced, more robust version of the snippet, while this would be a dead-simple example of rendering data to the DOM quickly and efficiently.
